### PR TITLE
feat(l2-contracts): OP-Stack L2 Solidity suite - SystemConfig slot-KV + L2ValidatorSet

### PR DESCRIPTION
**Pull request by kyonRay**
## What

OP-Stack L2-mode Solidity suite under `bcos-l2-contracts/`, gated behind the `WITH_L2_CONTRACTS` CMake option (default `OFF`) so a normal C++ node build does not require the Foundry toolchain. Two FISCO-written predeploys plus the scaffolding to pin and build the OP fork contracts.

### SystemConfig — `0x42000000000000000000000000000000000000C0`

Generic key→value config store (`setValueByKey` / `getValueByKey`), each entry packed as `(uint192 value, uint64 enableNumber)` in a single slot. The L2 upper layers (executor/consensus) read a key's slot directly without an EVM call, via `keccak256(utf8(key) ++ be32(baseSlot))`; the EVM getters are the external-caller path. New config keys need zero ABI change. Access control is OZ `OwnableUpgradeable` (owner = ProxyAdmin); slot addressing is locked to `keccak256`, decoupled from the node's SM3 hash in crypto-national mode.

### L2ValidatorSet — `0x42000000000000000000000000000000000000C1`

Single-record CRUD validator registry (`addValidator` / `removeValidator` / `updateValidator` / `getValidator` / `getValidators` / `isValidator`), all O(1), backed by OZ `EnumerableSet.AddressSet` plus a parallel record mapping. The `Validator` struct packs `feeAddress` + `jailed` + `votingPower` into one slot; `consensusPublicKey` is algorithm-neutral; the consensus address is the identity (set element + mapping key) and is not a struct field. `jailed` / `incoming` / `felony` / `misdemeanor` are Phase B placeholders.

### Build / vendoring

The OP fork is pinned (not vendored) in `op-fork-pin.toml`; OpenZeppelin is `forge install`-ed into a git-ignored `lib/` at the pinned SHAs. storage-layout golden fixtures back a CI drift gate.

## Test

```
cd bcos-l2-contracts
forge install foundry-rs/forge-std
forge install OpenZeppelin/openzeppelin-contracts@ecd2ca2cd7cac116f7a37d0e474bbb3d7d5e1c4d
forge install OpenZeppelin/openzeppelin-contracts-upgradeable@0a2cb9a445c365870ed7a8ab461b12acf3e27d63
forge build && forge test
```

24 tests green (SystemConfig 8 + L2ValidatorSet 16); `forge inspect <C> storage-layout` against the checked-in fixtures is clean.

## Notes

`WITH_L2_CONTRACTS` defaults `OFF`; this PR adds no dependency to the C++ node build. The C++ `L2ConfigLoader` that reads these predeploys lands separately.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bcos-l2-contracts/src/SystemConfig.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /// @notice Set the deploy-time owner (ProxyAdmin in production). Not run at
+    ///         genesis — see contract @dev note.
+    function initialize(address owner_) external initializer {
+        __Ownable_init();
+        _transferOwnership(owner_);
+    }
+
+    /// @inheritdoc ISystemConfig
+    function setValueByKey(string calldata key, uint192 value, uint64 enableNumber)
+        external
```
Comment (morebtcg) [inline bcos-l2-contracts/src/SystemConfig.sol:40]: 🔴 **`mapping(string => Entry) private _config` has no deletion mechanism.** If a malicious owner (ProxyAdmin compromised) fills this map, state-root bloat is unbounded. Add `removeKey(string calldata key)` (owner-only) or explicitly document the design decision to omit deletion.

Also: callers of `getValueByKey` must check `enableNumber <= block.number` themselves — document this in `ISystemConfig.sol` natspec.

**Review by morebtcg — COMMENTED** [review]
> ⚠️ 此 Review 已被中文版取代：https://github.com/FISCO-BCOS/FISCO-BCOS/pull/5271#pullrequestreview-4524211558

File: bcos-l2-contracts/src/L2ValidatorSet.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    ///         validator set. Not run at genesis — see contract @dev note.
+    function initialize(address[] memory addrs, Validator[] memory vals, address owner_)
+        external
+        initializer
+    {
+        require(addrs.length == vals.length, "length mismatch");
+        __Ownable_init();
+        _transferOwnership(owner_);
+        for (uint256 i = 0; i < addrs.length; i++) {
+            _add(addrs[i], vals[i]);
```
Comment (morebtcg) [inline bcos-l2-contracts/src/L2ValidatorSet.sol:44]: 🟡 **`Validator.incoming` (uint256, slot 2) is always 0 in Phase A.** If Phase B only needs `uint64`/`uint128`, it could be packed with `votingPower` in slot 0 to save one storage slot per validator. Add a comment documenting the Phase B plan for this field so future readers know whether the separate slot is intentional.

File: bcos-l2-contracts/foundry.toml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+src = "src"
+out = "out"
+libs = ["lib"]
+test = "test"
+solc = "0.8.25"
+optimizer = true
+optimizer_runs = 999999
+via_ir = false
+evm_version = "cancun"
+remappings = [
```
Comment (morebtcg) [inline bcos-l2-contracts/foundry.toml:19]: 🟡 **`optimizer_runs = 999999`** — correct for predeploys (deploy once, run infinitely), but verify all compiled bytecodes (including OP fork) stay under the 24KB Spurious Dragon limit. Add a CI bytecode-size check.

File: bcos-l2-contracts/src/SystemConfig.sol (hunk 1 above)
Comment (morebtcg) [inline bcos-l2-contracts/src/SystemConfig.sol:40]: 🔴 **`mapping(string => Entry) private _config` 没有删除机制。** 如果恶意 owner（ProxyAdmin 被攻陷）填充此映射，状态根膨胀无上限。请添加 `removeKey(string calldata key)` 函数（owner-only），或明确在文档中说明省略删除功能是有意为之的设计决策。

另外：`getValueByKey` 的调用者需要自行检查 `enableNumber <= block.number`——请在 `ISystemConfig.sol` 的 natspec 中文档化这一点。

File: bcos-l2-contracts/src/L2ValidatorSet.sol (hunk 2 above)
Comment (morebtcg) [inline bcos-l2-contracts/src/L2ValidatorSet.sol:44]: 🟡 **`Validator.incoming`（uint256，槽位 2）在 Phase A 中始终为 0。** 如果 Phase B 只需要 `uint64` 或 `uint128`，它可以与 `votingPower` 一起打包在槽位 0 中，每个验证者节省一个存储槽位。请在此字段上方添加注释，说明 Phase B 的计划，让未来的维护者知道独立槽位是有意为之。

File: bcos-l2-contracts/foundry.toml (hunk 3 above)
Comment (morebtcg) [inline bcos-l2-contracts/foundry.toml:19]: 🟡 **`optimizer_runs = 999999`** — 对预部署合约而言这是正确的（部署一次，无限运行），但请验证所有编译后的字节码（包括 OP fork 合约）均在 24KB Spurious Dragon 限制之内。请在 CI 中添加合约大小检查。

**Review by morebtcg — COMMENTED** [review]
## 审查总结

Solidity 合约设计良好 ✅。SystemConfig slot-KV 方案合理；L2ValidatorSet 的 O(1) CRUD 设计适合验证者集合。三个需要关注的问题：

**Review by morebtcg — APPROVED** [review]

**Review by ywy2090 — APPROVED** [review]
## Overview

PR #5271 introduces the OP-Stack L2 Solidity suite under `bcos-l2-contracts/`, gated behind `WITH_L2_CONTRACTS=OFF` (default) so normal C++ builds are unaffected. Two predeploys are implemented:

- **`SystemConfig`** (`0x42...00C0`): Generic key→value config store using `mapping(string => Entry)` with packed `(uint192 value, uint64 enableNumber)` per slot. The L2 C++ layers read slots directly via `keccak256(utf8(key) ‖ be32(baseSlot))` — no EVM call per block. Access control via OZ `OwnableUpgradeable`.
- **`L2ValidatorSet`** (`0x42...00C1`): O(1) CRUD validator registry backed by OZ `EnumerableSet.AddressSet` + parallel record mapping. `Validator` struct packs `feeAddress` + `jailed` + `votingPower` into one slot; `consensusPublicKey` is algorithm-neutral. Phase B placeholders (`felony`/`misdemeanor`) revert in Phase A.

The PR also includes CMake wiring, a Foundry project config (`foundry.toml`), an OP fork pin file (`op-fork-pin.toml`), and a detailed README with genesis deployment notes.

## Test coverage

24 tests across 2 suites, all green:
- **SystemConfig** (8 tests): set/get round-trip, unset→zero, onlyOwner guard, ConfigUpdate event emission, packing independence (max boundary values), overwrite semantics, re-initialize guard, and a cross-language slot-addressing test that verifies `keccak256(abi.encodePacked(key, CONFIG_BASE_SLOT))` against `vm.load` — pinning the C++ reader's storage formula.
- **L2ValidatorSet** (16 tests): init, get, add (onlyOwner + duplicate + zero address revert), remove (onlyOwner + swap-pop correctness + not-validator revert), update (onlyOwner + not-validator revert), getCurrentValidators addr/key pairing, struct packing boundary test, felony/misdemeanor Phase A revert, re-initialize guard, length mismatch revert.

## Verdict

Found: 0 critical, 0 important, 0 suggestions

**Approve** — clean, well-tested contracts with proper access control, storage layout verification, and deployment documentation.
